### PR TITLE
Implement metavariable matching between patterns

### PR DESCRIPTION
**Pull request by nbrahms**
This commit implements metavariable matching between patterns. What this
means:

  - Two patterns are equal only if, for any commonly defined
    metavariables, those metavariables refer to the same variable (if
    a variable) or expression
  - A pattern is "inside" another pattern only if, for any commonly defined
    metavariables, those metavariables refer to the same variable (if
    a variable) or expression

See semgrep/tests/e2e/rules/inside.yaml and
semgrep/tests/e2e/targets/basic/inside.py for an example of how this
metavariable matching works.

### CODE REVIEW

**Comment by nbrahms** [comment]
I'm guessing this will probably break some rules :(

File: semgrep/semgrep/semgrep_types.py [hunk 1]
```diff
@@ -118,9 +119,42 @@ def pattern_names_for_operators(operators: List[Operator]) -> List[str]:
...
-        return self.start <= other_range.start and other_range.end <= self.end
+        return (
+            self.start <= other_range.start
+            and other_range.end <= self.end
+            and self.vars_match(other_range)
+        )
+
+    def vars_match(self, rhs: "Range") -> bool:
+        """
+        Returns true if and only if all metavariables that match as variable nodes in either
```
Comment (nbrahms) [inline semgrep/semgrep/semgrep_types.py]: I'll reword this

File: semgrep/semgrep/semgrep_types.py [hunk 2]
```diff
@@ -118,9 +119,42 @@ def pattern_names_for_operators(operators: List[Operator]) -> List[str]:
 class Range(NamedTuple):
     start: int
     end: int
+    vars: Mapping[str, Any]
```
Comment (brendongo) [inline semgrep/semgrep/semgrep_types.py:122]: should this be a property of PatternMatch rather than Range?

Comment (nbrahms) [reply semgrep/semgrep/semgrep_types.py:122]: I need it for the equality and inclusion operations, so I'd have to call it all over the place if I did that. I think it makes more sense for an object to declare its own equality.

**Comment by nbrahms** [comment]
> I'm guessing this will probably break some rules :(

It didn't!

File: semgrep/semgrep/semgrep_types.py [hunk 3]
```diff
@@ -118,9 +119,42 @@ def pattern_names_for_operators(operators: List[Operator]) -> List[str]:
...
+        Returns true if and only if all metavariables in both this and the other Range refer to the same
+        variables (if variable nodes), in the same scope, or the same expressions (if expression nodes).
+
+        That is, if two patterns define a "$X", and $X refers to a variable in one pattern, then
+        $X must refer to the same variable in both patterns
+        :param rhs: The other Range
+        """
+        to_match = set(self.vars.keys()).intersection(rhs.vars.keys())
+        return all(
+            v in self.vars and v in rhs.vars and self.vars[v] == rhs.vars[v]
```
Comment (brendongo) [inline semgrep/semgrep/semgrep_types.py]: Do we need to do a membership check if we are looping over a set intersection?

**Review by brendongo — COMMENTED** [review]
looks good aside from membership check comment above.

File: semgrep/semgrep/semgrep_types.py (hunk 3 above)
Comment (nbrahms) [reply semgrep/semgrep/semgrep_types.py]: oh yeah that's stale

**Review by brendongo — APPROVED** [review]

**Comment by ievans** [comment]
@nbrahms does this address/close https://github.com/returntocorp/semgrep/issues/176?

**Pull request merged**